File: engine/src/scan/public-repo-scan.ts
import { assertPublicRepoAuditCap } from "../caps.js";
import { anchorPendingCertificatesAfterScan } from "../certificate-anchor.js";
import { ensureAuditCertificate } from "../audit-persistence.js";
import { getDb, nowIso } from "../db.js";
import { enqueueAuditJobs } from "../jobs/queue.js";
import type { LockfileDep } from "../lockfile/index.js";
import { getVerdict } from "../verdict-index.js";

export interface PublicRepoReference {
  owner: string;
  repo: string;
  fullName: string;
}

export class InvalidPublicRepoReferenceError extends Error {
  constructor() {
    super("Enter a GitHub repository as owner/repo or https://github.com/owner/repo");
    this.name = "InvalidPublicRepoReferenceError";
  }
}

const OWNER_PATTERN = /^[A-Za-z0-9](?:[A-Za-z0-9-]{0,38})$/;
const REPO_PATTERN = /^[A-Za-z0-9_.-]{1,100}$/;

/**
 * Accept only a GitHub repository identity, never an arbitrary fetch URL.
 * This is both the UX normalizer and the SSRF boundary for public audits.
 */
export function parsePublicRepoReference(input: string): PublicRepoReference {
  const value = input.trim();
  let path = value;

  if (/^github\.com\//i.test(path)) path = `https://${path}`;
  if (/^https?:\/\//i.test(path)) {
    let url: URL;
    try {
      url = new URL(path);
    } catch {
      throw new InvalidPublicRepoReferenceError();
    }
    if (
      url.protocol !== "https:" ||
      url.hostname.toLowerCase() !== "github.com" ||
      url.username ||
      url.password ||
      url.search ||
      url.hash
    ) {
      throw new InvalidPublicRepoReferenceError();
    }
    path = url.pathname.replace(/^\/+|\/+$/g, "");
  } else if (path.includes(":")) {
    throw new InvalidPublicRepoReferenceError();
  }

  const parts = path.replace(/^\/+|\/+$/g, "").split("/");
  if (parts.length !== 2) throw new InvalidPublicRepoReferenceError();

  const owner = parts[0] ?? "";
  const repo = (parts[1] ?? "").replace(/\.git$/i, "");
  if (!OWNER_PATTERN.test(owner) || !REPO_PATTERN.test(repo) || repo === "." || repo === "..") {
    throw new InvalidPublicRepoReferenceError();
  }
  return { owner, repo, fullName: `${owner}/${repo}` };
}

export interface CreatePublicRepoScanInput {
  installationId: number;
  requestedBy: number;
  githubRepoId: number;
  owner: string;
  name: string;
  fullName: string;
  htmlUrl: string;
  defaultBranch: string;
  commitSha: string | null;
  lockfilePath: string;
  lockfileSha: string;
  deps: LockfileDep[];
  accountLogin: string;
}

function uniqueDeps(deps: readonly LockfileDep[]): LockfileDep[] {
  return [...new Map(deps.map((dep) => [`${dep.name}\0${dep.version}`, dep])).values()];
}

export function findRunningPublicScan(
  installationId: number,
  fullName: string,
): { id: number } | null {
  return (
    (getDb()
      .prepare(
        `SELECT id FROM public_repo_scans
         WHERE installation_id = ? AND full_name = ? AND status = 'running'
         LIMIT 1`,
      )
      .get(installationId, fullName) as { id: number } | undefined) ?? null
  );
}

/** Create a read-only snapshot and enqueue only globally uncached work. */
export function createPublicRepoScan(input: CreatePublicRepoScanInput): number {
  const db = getDb();
  const deps = uniqueDeps(input.deps);
  const cachedKeys = new Set(
    deps.filter((dep) => getVerdict(dep.name, dep.version)).map((dep) => `${dep.name}\0${dep.version}`),
  );
  const misses = deps.filter((dep) => !cachedKeys.has(`${dep.name}\0${dep.version}`));

  let scanId = 0;
  db.transaction(() => {
    // A repository consumes one Free slot only once. The stable GitHub id
    // survives renames, and the transaction prevents concurrent new repos
    // from both claiming the final slot.
    assertPublicRepoAuditCap(input.installationId, input.githubRepoId);

    scanId = Number(
      db
        .prepare(
          `INSERT INTO public_repo_scans (
             installation_id, requested_by, github_repo_id, owner, name, full_name,
             html_url, default_branch, commit_sha, lockfile_path, lockfile_sha,
             status, total, cached
           ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'running', ?, ?)`,
        )
        .run(
          input.installationId,
          input.requestedBy,
          input.githubRepoId,
          input.owner,
          input.name,
          input.fullName,
          input.htmlUrl,
          input.defaultBranch,
          input.commitSha,
          input.lockfilePath,
          input.lockfileSha,
          deps.length,
          cachedKeys.size,
        ).lastInsertRowid,
    );

    const insertItem = db.prepare(
      `INSERT INTO public_repo_scan_items (scan_id, name, version, direct, range, cached)
       VALUES (?, ?, ?, ?, ?, ?)`,
    );
    for (const dep of deps) {
      insertItem.run(
        scanId,
        dep.name,
        dep.version,
        dep.direct ? 1 : 0,
        dep.range,
        cachedKeys.has(`${dep.name}\0${dep.version}`) ? 1 : 0,
      );
    }
  })();

  // Public snapshots do not own jobs: scan_id stays null so no installed repo,
  // check-run, or webhook relationship can be inferred for the target.
  enqueueAuditJobs(
    misses.map((dep) => ({
      packageName: dep.name,
      version: dep.version,
      org: input.accountLogin,
      scanId: null,
    })),
  );
  refreshPublicScanProgress(scanId);
  return scanId;
}

interface PublicScanItemState {
  name: string;
  version: string;
  cached: number;
  verdict: string | null;
  active: number;
}

function publicScanItemStates(scanId: number): PublicScanItemState[] {
  return getDb()
    .prepare(
      `SELECT psi.name, psi.version, psi.cached, pv.verdict,
              EXISTS(
                SELECT 1 FROM jobs j
                WHERE j.package_name = psi.name AND j.version = psi.version
                  AND j.state IN ('queued', 'running')
              ) AS active
       FROM public_repo_scan_items psi
       LEFT JOIN package_verdicts pv ON pv.name = psi.name AND pv.version = psi.version
       WHERE psi.scan_id = ?`,
    )
    .all(scanId) as PublicScanItemState[];
}

export function refreshPublicScanProgress(scanId: number): void {
  const db = getDb();
  const scan = db
    .prepare("SELECT status FROM public_repo_scans WHERE id = ?")
    .get(scanId) as { status: string } | undefined;
  if (!scan || scan.status !== "running") return;

  const items = publicScanItemStates(scanId);
  const cached = items.filter((item) => item.cached === 1).length;
  const audited = items.filter((item) => item.cached === 0 && item.verdict !== null).length;
  const active = items.filter((item) => item.active === 1).length;
  const failed = items.filter((item) => item.verdict === null && item.active === 0).length;

  if (active > 0) {
    db.prepare(
      `UPDATE public_repo_scans
       SET total = ?, cached = ?, audited = ?, failed = ? WHERE id = ?`,
    ).run(items.length, cached, audited, failed, scanId);
    return;
  }

  db.prepare(
    `UPDATE public_repo_scans
     SET status = 'done', total = ?, cached = ?, audited = ?, failed = ?, finished_at = ?
     WHERE id = ?`,
  ).run(items.length, cached, audited, failed, nowIso(), scanId);
  console.log(
    `[public-scan] #${scanId} done — ${items.length} items (${cached} cached, ${audited} audited, ${failed} unresolved)`,
  );

  let certificatesReady = 0;
  for (const item of items) {
    if (item.verdict === null) continue;
    if (ensureAuditCertificate(item.name, item.version)) {
      certificatesReady += 1;
    }
  }
  console.log(
    `[public-scan] #${scanId} prepared ${certificatesReady} certificate(s) for Merkle anchoring`,
  );
  anchorPendingCertificatesAfterScan();
}

export function refreshPublicScansTouching(packageName: string, version: string): void {
  const scans = getDb()
    .prepare(
      `SELECT DISTINCT prs.id FROM public_repo_scans prs
       JOIN public_repo_scan_items psi ON psi.scan_id = prs.id
       WHERE prs.status = 'running' AND psi.name = ? AND psi.version = ?`,
    )
    .all(packageName, version) as Array<{ id: number }>;
  for (const scan of scans) refreshPublicScanProgress(scan.id);
}

export interface PublicScanRollup {
  verdict: string | null;
  dangerous: number;
  suspect: number;
  unknown: number;
  safe: number;
}

export function computePublicScanRollup(scanId: number): PublicScanRollup {
  const rows = getDb()
    .prepare(
      `SELECT pv.verdict AS verdict, COUNT(*) AS c
       FROM public_repo_scan_items psi
       LEFT JOIN package_verdicts pv ON pv.name = psi.name AND pv.version = psi.version
       WHERE psi.scan_id = ?
       GROUP BY pv.verdict`,
    )
    .all(scanId) as Array<{ verdict: string | null; c: number }>;

  const rollup: PublicScanRollup = {
    verdict: null,
    dangerous: 0,
    suspect: 0,
    unknown: 0,
    safe: 0,
  };
  let total = 0;
  for (const row of rows) {
    total += row.c;
    if (row.verdict === "DANGEROUS") rollup.dangerous += row.c;
    else if (row.verdict === "SUSPECT") rollup.suspect += row.c;
    else if (row.verdict === "SAFE") rollup.safe += row.c;
    else rollup.unknown += row.c;
  }
  if (total === 0) return rollup;
  if (rollup.dangerous > 0) rollup.verdict = "DANGEROUS";
  else if (rollup.suspect > 0) rollup.verdict = "SUSPECT";
  else if (rollup.unknown > 0) rollup.verdict = "UNKNOWN";
  else rollup.verdict = "SAFE";
  return rollup;
}

File: engine/npmguard/panel/scan/public_repo_scan.py
"""Item discovery for the ``public_repo_scan`` origin: any public repo's lockfile.

A public-repo audit is a **read-only snapshot**: a signed-in user points the panel
at any public GitHub repository, and its root lockfile is audited against the
shared verdict cache. Unlike an owned-repo scan there is no repo row, no check run,
and no webhook relationship.

After R-1 the only things this module owns are the SSRF boundary, the cost
ceiling, and the snapshot subject row. Progress, the rollup, the item projection, truncation and the
SSE stream are the shared audit-set entity's — the second progress refresher and
the second rollup call site that used to live here are gone, and so is the
client-side polling loop they fed, because a public set streams on exactly the same
route as an owned-repo set.

Two boundaries remain load-bearing:

- :func:`parse_public_repo_reference` is the **SSRF boundary**. It accepts only a
  GitHub repository *identity* (``owner/repo`` or ``https://github.com/owner/repo``)
  — never an arbitrary fetch URL — and hands back a validated ``owner``/``repo``.
  The bytes are then pulled by the credential-free public octokit + the raw-host
  allow-list in ``github/content.py``; a private repo 404s by construction.
- The set is scoped by its **requester** (D-1): a signed-in user, no App
  installation, no ownership, nothing charged. ``audit_sets.requested_by`` is that
  identity — the key column of ``ix_audit_sets_active_public``, the read
  authorization, and the per-user budget in ``panel/public_limits.py``. The stable
  ``github_repo_id`` remains the set's ``origin_ref``, which is what replaced the
  lowercased ``full_name`` mirror the active-scan uniqueness used to need, and a
  rename still cannot smuggle in a second concurrent audit.

Billing (the one per-origin difference besides discovery): **nobody is billed**.
A public snapshot passes no budget hooks and no ``billed_to``; its ceiling is
cost, bounded per user, and a scan past that ceiling covers less of the lockfile
rather than being refused. What it covers versus what the lockfile held is
recorded on the snapshot row (``dep_count``) so the difference can be stated
instead of hidden.

Its jobs carry the REQUESTER's login as their queue fairness key, because
starving every other user behind one user's monorepo is a scheduling bug, not a
billing decision, and conflating the two is why these jobs were once ``org=None``.
TODO(R-2): they belong in the durable queue's ``public`` lane, which is where
"can never starve paid or panel work" (F-F6) is actually expressible — a lane
built here would be this repo's THIRD queue implementation, the exact mistake
R-1 was done to stop.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from urllib.parse import urlsplit

import sqlalchemy as sa
from sqlalchemy.ext.asyncio import async_sessionmaker

from ..audit_set import (
    ORIGIN_PUBLIC_REPO_SCAN,
    AuditSetSpec,
    AuditSetStore,
    dedupe,
)
from ..lockfile import LockfileDep
from ..public_limits import PublicScanLimits
from ..tables import audit_sets, gh_users, public_repo_scans

# GitHub identity grammar (mirrors the TS OWNER_PATTERN / REPO_PATTERN).
_OWNER_PATTERN = re.compile(r"^[A-Za-z0-9](?:[A-Za-z0-9-]{0,38})$")
_REPO_PATTERN = re.compile(r"^[A-Za-z0-9_.-]{1,100}$")
_GITHUB_HOST = "github.com"


class InvalidPublicRepoReferenceError(Exception):
    """The input was not a recognizable GitHub repository identity."""

    def __init__(self) -> None:
        super().__init__("Enter a GitHub repository as owner/repo or https://github.com/owner/repo")


@dataclass(frozen=True, slots=True)
class PublicRepoReference:
    owner: str
    repo: str
    full_name: str


def parse_public_repo_reference(raw: str) -> PublicRepoReference:
    """Accept only a GitHub repository identity, never an arbitrary fetch URL.

    This is both the UX normalizer and the SSRF boundary for public audits:
    ``owner/repo`` or an ``https://github.com/owner/repo`` URL (``.git`` stripped)
    is accepted; anything with another scheme/host, credentials, a query, a
    fragment, or a shape that is not exactly two path segments is rejected.
    """
    value = raw.strip()
    path = value

    # A bare ``github.com/owner/repo`` (no scheme) is normalized to https first.
    if re.match(r"^github\.com/", path, re.IGNORECASE):
        path = f"https://{path}"

    if re.match(r"^https?://", path, re.IGNORECASE):
        parts = urlsplit(path)
        host = (parts.hostname or "").lower()
        if (
            parts.scheme != "https"
            or host != _GITHUB_HOST
            or parts.username
            or parts.password
            or parts.query
            or parts.fragment
        ):
            raise InvalidPublicRepoReferenceError()
        path = parts.path.strip("/")
    elif ":" in path:
        # Reject scp-style (``git@github.com:owner/repo``) and any other scheme —
        # a colon in a non-URL input is never a valid ``owner/repo``.
        raise InvalidPublicRepoReferenceError()

    segments = path.strip("/").split("/")
    if len(segments) != 2:
        raise InvalidPublicRepoReferenceError()

    owner = segments[0]
    repo = re.sub(r"\.git$", "", segments[1], flags=re.IGNORECASE)
    if not _OWNER_PATTERN.match(owner) or not _REPO_PATTERN.match(repo) or repo in {".", ".."}:
        raise InvalidPublicRepoReferenceError()
    return PublicRepoReference(owner=owner, repo=repo, full_name=f"{owner}/{repo}")


@dataclass(frozen=True, slots=True)
class CreatePublicRepoScanInput:
    """Everything needed to persist one public-repo snapshot.

    ``deps`` is the parsed lockfile; ``github_repo_id`` is the set's
    ``origin_ref``; the ``owner``/``name``/``full_name`` are the canonical values
    from the GitHub repo response, not the user's input.
    """

    requested_by: int
    github_repo_id: int
    owner: str
    name: str
    full_name: str
    html_url: str
    default_branch: str
    commit_sha: str | None
    lockfile_path: str
    lockfile_sha: str
    deps: list[LockfileDep]


@dataclass
class PublicRepoScanEngine:
    """Discovers a public repo's items, opens the set, writes the snapshot row."""

    sessions: async_sessionmaker
    limits: PublicScanLimits
    sets: AuditSetStore

    # -- lookups -----------------------------------------------------------

    async def find_running_public_scan(self, user_id: int, github_repo_id: int) -> int | None:
        """The set id of a still-live audit of this repo for this USER, or ``None``.

        Keyed on the stable ``github_repo_id``, so a rename cannot smuggle in a
        second concurrent audit — the case-insensitive ``full_name`` mirror this
        replaced could. The frontend treats a 409 carrying this id as a success:
        the set is already live and streamable.
        """
        async with self.sessions() as session:
            return (
                await session.execute(
                    sa.select(audit_sets.c.id)
                    .where(
                        audit_sets.c.origin == ORIGIN_PUBLIC_REPO_SCAN,
                        audit_sets.c.origin_ref == github_repo_id,
                        audit_sets.c.requested_by == user_id,
                        audit_sets.c.finished_at.is_(None),
                    )
                    .limit(1)
                )
            ).scalar_one_or_none()

    # -- set creation ------------------------------------------------------

    async def create_public_repo_scan(self, data: CreatePublicRepoScanInput) -> int:
        """Open a ``public_repo_scan`` set and persist its snapshot subject row.

        The requester's live-scan concurrency is re-asserted here as the race
        guard, and the ``ix_audit_sets_active_public`` partial unique index is the
        durable guard against two concurrent live audits of one repo by one user.

        The set covers the lockfile's cached verdicts plus at most
        ``new_audit_budget`` misses, so a 900-dep monorepo cannot buy 900 audits.
        ``dep_count`` records what the lockfile actually held, which is the only
        way the difference can be stated later — the set knows what it covers and
        nothing about what it declined to.

        Returns the SET id — which is also the snapshot's id, because ``set_id`` is
        the snapshot's primary key. One id, so ``scanId`` means the same thing on
        every route.
        """
        await self.limits.assert_capacity(data.requested_by)
        set_id = await self.sets.create(
            AuditSetSpec(
                origin=ORIGIN_PUBLIC_REPO_SCAN,
                origin_ref=data.github_repo_id,
                trigger="manual",
                items=data.deps,
                requested_by=data.requested_by,
                # The queue fairness key is the REQUESTER's login, not an org.
                # `billed_org` is not a billing field (money is metered by the
                # budget hooks, which this origin passes none of) — it is what
                # `claim_next` groups on, and leaving it NULL would drop every
                # public scan into one bucket, so one user's 900-dep monorepo
                # would starve every other user's scan.
                billed_org=await self._login_of(data.requested_by),
                max_new_audits=await self.limits.new_audit_budget(data.requested_by),
                commit_sha=data.commit_sha,
            )
        )
        async with self.sessions() as session, session.begin():
            await session.execute(
                public_repo_scans.insert().values(
                    set_id=set_id,
                    github_repo_id=data.github_repo_id,
                    owner=data.owner,
                    name=data.name,
                    full_name=data.full_name,
                    html_url=data.html_url,
                    default_branch=data.default_branch,
                    lockfile_path=data.lockfile_path,
                    lockfile_sha=data.lockfile_sha,
                    dep_count=len(dedupe(data.deps)),
                )
            )
        return set_id

    async def _login_of(self, user_id: int) -> str | None:
        async with self.sessions() as session:
            return (
                await session.execute(sa.select(gh_users.c.login).where(gh_users.c.id == user_id))
            ).scalar_one_or_none()


__all__ = [
    "CreatePublicRepoScanInput",
    "InvalidPublicRepoReferenceError",
    "PublicRepoReference",
    "PublicRepoScanEngine",
    "parse_public_repo_reference",
]
